packages/azureblobstorage/src/Adaptor.js

```diff
@@ -0,0 +1,216 @@
+import {
+  execute as commonExecute,
+  composeNextState,
+} from '@openfn/language-common';
+import { BlobServiceClient, StorageSharedKeyCredential } from '@azure/storage-blob';
+
+
+/**
+ * Execute a sequence of operations.
+ * Wraps `language-common/execute` to make working with this API easier.
+ * @example
+ * execute(
+ *   create('foo'),
+ *   delete('bar')
+ * )(state)
+ * @private
+ * @param {Operations} operations - Operations to be performed.
+ * @returns {Operation}
+ */
+export function execute(...operations) {
+  const initialState = {
+    references: [],
+    data: null,
+  };
+
+  return (state) => {
+    return commonExecute(
+      createClient,
+      ...operations,
+      teardownClient
+    )({ ...initialState, ...state });
+  };
+}
+
+// Create client for Azure Blob Storage
+function createClient(state) {
+  const { accountName, accountKey } = state.configuration;
+  const blobStorageUrl = `https://${accountName}.blob.core.windows.net`;
+  const sharedKeyCredential = new StorageSharedKeyCredential(accountName, accountKey);
+  state.client = new BlobServiceClient(blobStorageUrl, sharedKeyCredential);
+  return state;
+}
+
+// Destroy client
+function teardownClient(state) {
+  if (state.client) {
+    // Perform any other cleanup or disconnection logic if required
+    delete state.client;
+  }
+  return state;
+}
+
+// A basic Azure BlobServiceClient client is attached to state.client
+// but almost all operations require a ContainerClient. This function
+// just keeps things a little DRYer.
+function getContainerClient(state, containerName) {
+  const { client } = state;
+  return client.getContainerClient(containerName);
+}
+
+// DRY helper function, similar to getContainerClient, above,
+// but returns a BlockBlobClient
+function getBlockBlobClient(state, containerName, blobName) {
+  const containerClient = getContainerClient(state, containerName);
+  return containerClient.getBlockBlobClient(blobName);
+}
+
+/**
+ * Create a new container in Azure Blob Storage.
+ * @public
+ * @example
+ * createContainer('mycontainer')
+ * @function
+ * @param {string} containerName - Name of the container to create.
+ * @returns {Operation}
+ */
+export function createContainer(containerName) {
+  return async (state) => {
+    const containerClient = getContainerClient(state, containerName);
+    console.info(`Creating container '${containerName}'`);
+    const response = await containerClient.create();
+    console.debug(`Container '${containerName}' created successfully.`);
+    const nextState = composeNextState(state, response);
+    return nextState;
+  };
+}
+
+
+/**
+ * Check if a blob exists in Azure Blob Storage.
+ * @public
+ * @example
+ * checkBlobExists('mycontainer', 'myblob.txt')
+ * @function
+ * @param {string} containerName - Name of the container containing the blob.
+ * @param {string} blobName - Name of the blob to check.
+ * @returns {Operation}
+ */
+export function checkBlobExists(containerName, blobName) {
```

Comment:
My main thought is that there may be circumstances when people don't want to clobber an existing blob. Given this, it probably makes sense to change the `uploadBlob` signature to let the user control the overwrite behavior  (`overwrite = true | false`) and eliminate this public function.

Reply:
Yep -  a flag on the upload looks like a better solution to me!

You'll need to decider whether to throw and kill the job, or to return a status to state or via callback. I presume that if the write fails the.jon writer will want to change the id or something, rather than abandon the job 🤔

Reply:
I need to think more about this. For the moment, I lean towards killing the job and handling exceptions through a failure edge.

Reply:
I think the pattern you have now is great - let's use it as-is and see how it feels "in the wild".